### PR TITLE
breaks loop in FindPredecessor

### PR DESCRIPTION
**Pull request by msmedes**
Veni, Vidi, Hacki

### CODE REVIEW

**Comment by msmedes** [comment]
I deactivated that test so I can merge this after your review so you can mess with it if I don't figure it out

**Comment by msmedes** [comment]
PS to help debug GRPC issues set the following two environmental variables in your ~/.zshrc or ~./bashrc

```
export GRPC_GO_LOG_VERBOSITY_LEVEL=99
export GRPC_GO_LOG_SEVERITY_LEVEL=warning
```

**Comment by msmedes** [comment]
I think we need to create some kind of variadic functions to handle the following case:
(from `FindPredecessor`)
```		
if keyspace.Equal(n1.GetID(), node.id) {
	n1, err = node.ClosestPrecedingFinger(key)
} else {
	n1, err = n1.ClosestPrecedingFinger(key)
}
if err != nil {
	return nil, err
}
```
In `FindPredecessor` alone that boilerplate is in there 3 times, missing one of them was leading to the tests failing before.  I experimented with a variadic function (ie passing `node.ClosestPrecedingFinger` and `n1.ClosestPrecedingFinger` as variables, along with and args interface), and I'm not sure it removed much code or improved the readability of the code.  I'll mess around with it a little more.

**Comment by msmedes** [comment]
ok probably a bad idea.  we'd have to do something like this:
```
type function func(...interface{}) (scale.RemoteNode, error)

type LocalRemoteFunctions struct {
	nodeFunction   function
	remoteFunction function
}
```
and 
```
func RunLocalOrRemote(nodeID, remoteID scale.Key, functions *LocalRemoteFunctions, args []interface{}) (scale.RemoteNode, error) {
	var remoteNode scale.RemoteNode
	var err error

	if keyspace.Equal(nodeID, remoteID) {
		remoteNode, err = functions.nodeFunction(args...)
	} else {
		remoteNode, err = functions.remoteFunction(args...)
	}
	if err != nil {
		return nil, err
	}
	return remoteNode, nil
}
```
problem is, none of the functions take `...[]interface{}` as a parameter, and I could swap them all over to taking an `interface`, but then we'd lose all of our type checking, which seems like a bad idea. (obviously we could unpack the interfaces and manually check to see if all the variables passed are the correct type but no thanks).  I'll see if I can swing anything else.

File: internal/pkg/node/node.go [hunk 1]
```diff
@@ -436,34 +470,43 @@ func (node *Node) checkPredecessor() {
...
 		node.predecessor = nil
+		predecessor.CloseConnection()
+		delete(remotes.data, predecessor.GetAddr())
 	}
 }
 
 // Notify is called when another node thinks it is our predecessor
 func (node *Node) Notify(id scale.Key, addr string) error {
-	node.mutex.Lock()
-	defer node.mutex.Unlock()
```
Comment (msmedes) [inline internal/pkg/node/node.go:480]: removed the Lock() defer Unlock() because when `node.bootstrap()` was called it caused a deadlock.

Comment (thejchap) [reply internal/pkg/node/node.go:480]: ayyyy fun times

**Review by thejchap — APPROVED** [review]

**Comment by thejchap** [comment]
agreed on it being a good idea to clean up how things are being invoked locally or remotely. will take a look at the hax this evening!

**Pull request merged**